error_writer.go

```diff
@@ -41,86 +41,58 @@ const (
 		requireConnectProtocolHeader: config.RequireConnectProtocolHeader,
 	}
-	for name := range config.Codecs {
-		unary := connectContentTypeFromCodecName(StreamTypeUnary, name)
-		writer.unaryConnectContentTypes[unary] = struct{}{}
-		streaming := connectContentTypeFromCodecName(StreamTypeBidi, name)
-		writer.streamingConnectContentTypes[streaming] = struct{}{}
-	}
-	if config.HandleGRPC {
-		writer.grpcContentTypes[grpcContentTypeDefault] = struct{}{}
-		for name := range config.Codecs {
-			ct := grpcContentTypeFromCodecName(false /* web */, name)
-			writer.grpcContentTypes[ct] = struct{}{}
-		}
-	}
-	if config.HandleGRPCWeb {
-		writer.grpcWebContentTypes[grpcWebContentTypeDefault] = struct{}{}
-		for name := range config.Codecs {
-			ct := grpcContentTypeFromCodecName(true /* web */, name)
-			writer.grpcWebContentTypes[ct] = struct{}{}
-		}
-	}
-	return writer
 }
 
 func (w *ErrorWriter) classifyRequest(request *http.Request) protocolType {
 	ctype := canonicalizeContentType(getHeaderCanonical(request.Header, headerContentType))
-	if _, ok := w.unaryConnectContentTypes[ctype]; ok {
-		if err := connectCheckProtocolVersion(request, w.requireConnectProtocolHeader); err != nil {
-			return unknownProtocol
-		}
-		return connectUnaryProtocol
-	}
-	if _, ok := w.streamingConnectContentTypes[ctype]; ok {
+	isPost := request.Method == http.MethodPost
+	isGet := request.Method == http.MethodGet
+	switch {
+	case w.handleGRPC && isPost && (ctype == grpcContentTypeDefault || strings.HasPrefix(ctype, grpcContentTypePrefix)):
+		return grpcProtocol
+	case w.handleGRPCWeb && isPost && (ctype == grpcWebContentTypeDefault || strings.HasPrefix(ctype, grpcWebContentTypePrefix)):
+		return grpcWebProtocol
+	case isPost && strings.HasPrefix(ctype, connectStreamingContentTypePrefix):
 		// Streaming ignores the requireConnectProtocolHeader option as the
 		// Content-Type is enough to determine the protocol.
 		if err := connectCheckProtocolVersion(request, false /* required */); err != nil {
 			return unknownProtocol
 		}
 		return connectStreamProtocol
-	}
-	if _, ok := w.grpcContentTypes[ctype]; ok {
-		return grpcProtocol
-	}
-	if _, ok := w.grpcWebContentTypes[ctype]; ok {
-		return grpcWebProtocol
-	}
-	// Check for Connect-Protocol-Version header or connect protocol query
-	// parameter to support connect GET requests.
-	if request.Method == http.MethodGet {
-		connectVersion := getHeaderCanonical(request.Header, connectProtocolVersion)
-		if connectVersion == connectProtocolVersion {
-			return connectUnaryProtocol
+	case isPost && strings.HasPrefix(ctype, connectUnaryContentTypePrefix):
+		if err := connectCheckProtocolVersion(request, w.requireConnectProtocolHeader); err != nil {
+			return unknownProtocol
 		}
-		connectVersion = request.URL.Query().Get(connectUnaryConnectQueryParameter)
-		if connectVersion == connectUnaryConnectQueryValue {
-			return connectUnaryProtocol
+		return connectUnaryProtocol
+	case isGet && ctype == "":
```

Comment:
In order to be compatible with versions of connect-go prior to v1.13 ([which includes a fix for the content-type with GET requests](https://github.com/connectrpc/connect-go/pull/644/files#diff-cdd39981f92fb3d0b9c81ce742c64f5574d2aed55233f5fdd21a4a7ee80677f3R1049)), perhaps we should be more lenient here and either disregard the content-type or allow it to have a `connectUnaryContentTypePrefix` prefix.

Reply:
Makes sense, dropped the check.